lib/script.ts

```diff
@@ -32,28 +31,97 @@ export default class Script {
+    } else if (container.isCluster) {
+      return usingCluster(this, container, args, options, callback);
+    } else {
+      return usingStandalone(this, container, args, options, callback);
     }
+  }
+}
+
+function createEvalShaWithFallback(script, args, options, client) {
+  const evalsha = createEvalSha(script, args, options);
+  evalsha.promise = evalsha.promise.catch((err: Error) => {
+    if (err.toString().indexOf("NOSCRIPT") === -1) {
+      throw err;
+    }
+
+    // Do an eval as fallback, redis will hash and load it
+    const evalcmd = new Command("eval", [script.lua].concat(args), options);
+    evalcmd.promise = evalcmd.promise.then((r) => {
+      client._addedScriptHashes[script.sha] = true;
+      return r;
+    });
+    return client.sendCommand(evalcmd);
+  });
+  return evalsha;
+}
+
+function createEvalSha(script, args, options) {
+  const evalsha = new Command("evalsha", [script.sha].concat(args), options);
+  evalsha.isCustomCommand = true;
+  return evalsha;
+}
 
-    // result is not a Promise--probably returned from a pipeline chain; however,
-    // we still need the callback to fire when the script is evaluated
+// Pipeline mode (cluster and regular)
+function usingPipeline(script, pipeline, args, options, callback) {
+  // The script was loaded explicitly in this pipeline,
+  // so we can directly execute evalsha without loading the script again
+  if (pipeline._addedScriptHashes && pipeline._addedScriptHashes[script.sha]) {
+    const evalsha = createEvalSha(script, args, options);
     asCallback(evalsha.promise, callback);
+    return pipeline.sendCommand(evalsha);
+  }
 
-    return result;
+  // The script is loaded in redis already, so we try an evalsha
+  // and fallback to loading the script if it fails
+  if (pipeline.redis._addedScriptHashes[script.sha]) {
+    const evalsha = createEvalShaWithFallback(
+      script,
+      args,
+      options,
+      pipeline.redis
+    );
+    asCallback(evalsha.promise, callback);
+    return pipeline.sendCommand(evalsha);
   }
+
+  // If the script is not present on redis or in the pipeline,
+  // we use eval to load the script into the pipeline
+  if (!pipeline._addedScriptHashes) pipeline._addedScriptHashes = {};
+  pipeline._addedScriptHashes[script.sha] = true;
+
+  const evalcmd = new Command("eval", [script.lua].concat(args), options);
+  evalcmd.promise = evalcmd.promise.then((r) => {
+    pipeline.redis._addedScriptHashes[script.sha] = true;
+    return r;
+  });
+  asCallback(evalcmd.promise, callback);
+  return pipeline.sendCommand(evalcmd);
+}
+
+// Standalone mode (cluster)
+function usingCluster(script, cluster, args, options, callback) {
+  const evalsha = createEvalShaWithFallback(script, args, options, cluster);
+  asCallback(evalsha.promise, callback);
+  return cluster.sendCommand(evalsha);
+}
+
+// Standalone mode (regular)
+function usingStandalone(script, redis, args, options, callback) {
+  if (redis._addedScriptHashes[script.sha]) {
```

Comment:
Instead of doing those decisions in here, we could move it into the `command.toWritable` function.
But we don't have any stream context there. https://github.com/luin/ioredis/blob/b8177479c348aa4bbd467fa944d61fe9b35aec19/lib/redis/index.ts#L790